spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,301 @@
+      this.endOffset = calculateEndOffset(startOffset);
+    }
+  }
+
+  @Override
+  public Offset getStartOffset() {
+    if (startOffset == null) {
+      throw new IllegalStateException("Start offset is not set");
+    }
+
+    return startOffset;
+  }
+
+  @Override
+  public Offset getEndOffset() {
+    if (endOffset == null) {
+      throw new IllegalStateException("End offset is not set");
+    }
+
+    return endOffset;
+  }
+
+  @Override
+  public Offset deserializeOffset(String json) {
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {
+    // Since all the data and metadata of Iceberg is as it is, nothing needs to commit when
+    // offset is processed, so no need to implement this method.
+  }
+
+  @Override
+  public void stop() {}
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected List<CombinedScanTask> tasks() {
+    if (startOffset.equals(endOffset)) {
+      LOG.info("Start offset {} equals to end offset {}, no data to process", startOffset, endOffset);
+      return Collections.emptyList();
+    }
+
+    Preconditions.checkState(pendingTasks != null,
+        "pendingTasks is null, which is unexpected as it will be set when calculating end offset");
+
+    if (Iterables.isEmpty(pendingTasks)) {
+      LOG.info("There's no task to process in this batch");
+      return Collections.emptyList();
+    }
+
+    IndexedTask last = Iterables.getLast(pendingTasks);
+    if (last.snapshotId() != endOffset.snapshotId() || last.index() != endOffset.index()) {
+      throw new IllegalStateException("The cached pendingTasks doesn't match the current end offset " + endOffset);
+    }
+
+    LOG.info("Processing data from {} exclusive to {} inclusive", startOffset, endOffset);
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
+        CloseableIterable.transform(pendingTasks, IndexedTask::task), splitSize());
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize(), splitLookback(), splitOpenFileCost()));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, -1, true);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), -1, true);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    this.pendingTasks = getChangesWithRateLimit(start.snapshotId(), start.index(),
+        start.isStartingSnapshotId(), maxFilesPerBatch);
```

Comment:
If `getChangesWithRateLimit` is called every batch, then this is potentially going to plan a full table scan every batch.

I think we need to separate planning splits from consuming them. This method should refresh the table to pull in the latest changes, then consume N files from an existing iterator of file tasks. If there aren't enough files, it stops there. Then tasks returns the list of tasks that this consumed.

It's important that the iterator is independent of this call so that this doesn't plan every time. The current approach works at small scales, but if you want to start with an existing table this would be a problem.

Reply:
Can we build a utility iterator that would allow us to stream files for a given snapshot incrementally without loading all snapshot metadata at any given point in time? It seems it should be enough to load at most one manifest at a time. Or a couple of them if that improves performance.

For example, the iterator can accept a snapshot or a list of manifest files and a starting file index. The order of manifests should be stable (either with a manifest list or with manifests directly). If not, we can sort them by path or any other attribute during initialization. Then we can skip manifests based on our starting file index by looking at file stats for each manifest. For example, if our starting file index is 1000 and the first manifest has 100 existing and 200 added files, we know we can skip it as it contains only 300 live files. This will allow us to efficiently locate the first manifest to start with.

The state of the iterator can be represented as something like below:

```
List<ManifestFile> manifests;
int manifestIndex; // index of the manifest that is being processed
Iterable<DataFile> dataFiles; // live files for the manifest that is being processed
```

We can offer methods `next(numFiles)`, `hasNext()`, `next()` that would return `DataFile`s. When we can consume from that iterator in the streaming reader until we have enough files for a micro-batch.

Reply:
If the mentioned logic is not flexible enough to handle retries, we can modify it but the main idea is to rely on the manifest order and stats for live files.